File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-3.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        dirs = [(-1, -1), (0, -1), (1, -1), (1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0)]
        queue = deque()
        visited = set()

        ROWS, COLS = len(grid), len(grid[0])

        if not grid[0][0]:
            queue.append((0, 0, 1))
            visited.add((0, 0))

        while queue:
            qlen = len(queue)
            for _ in range(qlen):
                r, c, l = queue.popleft()

                for dr, dc in dirs:
                    nr, nc = r + dr, c + dc
                    if min(nr, nc) >= 0 and nr < ROWS and nc < COLS and not grid[nr][nc] and (nr, nc) not in visited:
                        queue.append((nr, nc, l + 1))
                        visited.add((nr, nc))
                        if nr == ROWS - 1 and nc == COLS - 1:
                            return l + 1
        return -1

File: Data Structures & Algorithms/count-number-of-islands/submission-3.cpp
class Solution {
    int ROWS, COLS;
    vector<pair<int, int>> dirs = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
public:
    void dfs(vector<vector<char>>& grid, int row, int col) {
        grid[row][col] = '0';
        for (pair<int, int> dir: dirs) {
            int nr = dir.first + row, nc = dir.second + col;
            if (min(nr, nc) < 0 || nr >= ROWS || nc >= COLS || grid[nr][nc] != '1') continue;
            dfs(grid, nr, nc);
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        int res = 0;
        ROWS = grid.size();
        COLS = grid[0].size();
        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == '1') {
                    dfs(grid, r, c);
                    res++;
                }
            }
        }
        return res;
    }
};
